audio/src/analyser_node.rs

```diff
@@ -0,0 +1,274 @@
+        self.max_decibels = max_decibels;
+    }
+
+    pub fn get_max_decibels(&self) -> f64 {
+        self.max_decibels
+    }
+
+    fn advance(&mut self) {
+        self.current_block += 1;
+        if self.current_block >= MAX_BLOCK_COUNT {
+            self.current_block = 0;
+        }
+    }
+
+    /// Get the data of the current block
+    fn curent_block_mut(&mut self) -> &mut [f32] {
+        let index = FRAMES_PER_BLOCK_USIZE * self.current_block;
+        &mut self.data[index..(index + FRAMES_PER_BLOCK_USIZE)]
+    }
+
+    /// Given an index from 0 to fft_size, convert it into an index into
+    /// the backing array
+    fn convert_index(&self, index: usize) -> usize {
+        let offset = self.fft_size - index;
+        let last_element = (1 + self.current_block) * FRAMES_PER_BLOCK_USIZE - 1;
+        if offset > last_element {
+            MAX_FFT_SIZE - offset + last_element
+        } else {
+            last_element - offset
+        }
+    }
+
+    /// Given an index into the backing array, increment it
+    fn advance_index(&self, index: &mut usize) {
+        *index += 1;
+        if *index >= MAX_FFT_SIZE {
+            *index = 0;
+        }
+    }
+
+    pub fn push(&mut self, mut block: Block) {
+        debug_assert!(block.chan_count() == 1);
+        self.advance();
+        if !block.is_silence() {
+            self.curent_block_mut().copy_from_slice(block.data_mut());
+        }
+        self.fft_computed = false;
+    }
+
+    /// https://webaudio.github.io/web-audio-api/#blackman-window
+    fn compute_blackman_windows(&mut self) {
+        if self.blackman_windows.len() == self.fft_size {
+            return;
+        }
+        const ALPHA: f32 = 0.16;
+        const ALPHA_0: f32 = (1. - ALPHA) / 2.;
+        const ALPHA_1: f32 = 1. / 2.;
+        const ALPHA_2: f32 = ALPHA / 2.;
+        self.blackman_windows.resize(self.fft_size, 0.);
+        let coeff = PI * 2. / self.fft_size as f32;
+        for n in 0..self.fft_size {
+            self.blackman_windows[n] = ALPHA_0 - ALPHA_1 * (coeff * n as f32).cos() 
+                                               + ALPHA_2 * (2. * coeff * n as f32).cos();
+        }
+    }
+
+    fn apply_blackman_window(&mut self) {
+        self.compute_blackman_windows();
+        self.windowed.resize(self.fft_size, 0.);
+
+        let mut data_idx = self.convert_index(0);
+        for n in 0..self.fft_size {
+            self.windowed[n] = self.blackman_windows[n] * self.data[data_idx];
+            self.advance_index(&mut data_idx);
+        }
+        println!("windowed {:?}", self.windowed);
+    }
+
+    fn compute_fft(&mut self) {
+        if self.fft_computed {
+            return;
+        }
+        self.fft_computed = true;
+        self.apply_blackman_window();
+        self.computed_fft_data.resize(self.fft_size / 2, 0.);
+        self.smoothed_fft_data.resize(self.fft_size / 2, 0.);
+
+        for k in 0..(self.fft_size / 2) {
```

Comment:
`(self.fft_size / 2) - 1` ?

Reply:
Nope, Rust ranges are open on the other end. `0..n` means "zero to n including zero but not n". There's `0..=n` for ranges closed on both ends.